### ISSUE TITLE
NewLaunch is compiled with Java 8, causing issues with legacy Forge

### ISSUE THREAD

**Issue by exaskye** [PolyMC/PolyMC#515]
### Operating System

Linux

### Description of bug

When launching an old version of Forge that uses an old ASM (e.g. 1.4.7, which uses ASM 4.0) that does not recognize class file version 52, an error will be printed to console:
```
2022-05-01 01:46:49 [SEVERE] [ForgeModLoader] Unable to read a class file correctly
java.lang.IllegalArgumentException
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at cpw.mods.fml.common.discovery.asm.ASMModParser.<init>(ASMModParser.java:41)
	at cpw.mods.fml.common.discovery.JarDiscoverer.discover(JarDiscoverer.java:55)
	at cpw.mods.fml.common.discovery.ContainerType.findMods(ContainerType.java:30)
	at cpw.mods.fml.common.discovery.ModCandidate.explore(ModCandidate.java:52)
	at cpw.mods.fml.common.discovery.ModDiscoverer.identifyMods(ModDiscoverer.java:103)
	at cpw.mods.fml.common.Loader.identifyMods(Loader.java:338)
	at cpw.mods.fml.common.Loader.loadMods(Loader.java:470)
	at cpw.mods.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:161)
	at net.minecraft.client.Minecraft.a(Minecraft.java:412)
	at asq.a(SourceFile:56)
	at net.minecraft.client.Minecraft.run(Minecraft.java:746)
	at java.lang.Thread.run(Thread.java:750)
2022-05-01 01:46:49 [SEVERE] [ForgeModLoader] There was a problem reading the entry net/minecraft/Launcher.class in the jar /usr/share/jars/NewLaunch.jar - probably a corrupt zip
cpw.mods.fml.common.LoaderException: java.lang.IllegalArgumentException
	at cpw.mods.fml.common.discovery.asm.ASMModParser.<init>(ASMModParser.java:47)
	at cpw.mods.fml.common.discovery.JarDiscoverer.discover(JarDiscoverer.java:55)
	at cpw.mods.fml.common.discovery.ContainerType.findMods(ContainerType.java:30)
	at cpw.mods.fml.common.discovery.ModCandidate.explore(ModCandidate.java:52)
	at cpw.mods.fml.common.discovery.ModDiscoverer.identifyMods(ModDiscoverer.java:103)
	at cpw.mods.fml.common.Loader.identifyMods(Loader.java:338)
	at cpw.mods.fml.common.Loader.loadMods(Loader.java:470)
	at cpw.mods.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:161)
	at net.minecraft.client.Minecraft.a(Minecraft.java:412)
	at asq.a(SourceFile:56)
	at net.minecraft.client.Minecraft.run(Minecraft.java:746)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.IllegalArgumentException
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at cpw.mods.fml.common.discovery.asm.ASMModParser.<init>(ASMModParser.java:41)
	... 11 more
2022-05-01 01:46:49 [WARNING] [ForgeModLoader] Zip file NewLaunch.jar failed to read properly, it will be ignored
cpw.mods.fml.common.LoaderException: java.lang.IllegalArgumentException
	at cpw.mods.fml.common.discovery.asm.ASMModParser.<init>(ASMModParser.java:47)
	at cpw.mods.fml.common.discovery.JarDiscoverer.discover(JarDiscoverer.java:55)
	at cpw.mods.fml.common.discovery.ContainerType.findMods(ContainerType.java:30)
	at cpw.mods.fml.common.discovery.ModCandidate.explore(ModCandidate.java:52)
	at cpw.mods.fml.common.discovery.ModDiscoverer.identifyMods(ModDiscoverer.java:103)
	at cpw.mods.fml.common.Loader.identifyMods(Loader.java:338)
	at cpw.mods.fml.common.Loader.loadMods(Loader.java:470)
	at cpw.mods.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:161)
	at net.minecraft.client.Minecraft.a(Minecraft.java:412)
	at asq.a(SourceFile:56)
	at net.minecraft.client.Minecraft.run(Minecraft.java:746)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.IllegalArgumentException
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at org.objectweb.asm.ClassReader.<init>(Unknown Source)
	at cpw.mods.fml.common.discovery.asm.ASMModParser.<init>(ASMModParser.java:41)
	... 11 more
```

This is ultimately harmless, as NewLaunch doesn't need to be used in the relaunched FML class loader, but it's a really large and unsightly error, and if something were to try to access the classes in NewLaunch from within the game, it would fail as FML has excluded it from the classpath.

This does not occur in MultiMC, as they compile NewLaunch for Java 7.

### Steps to reproduce

Install Forge in a 1.4.7 instance and launch the game.

### Suspected cause

NewLaunch is being compiled with a target of 8. It needs to be compiled with a target of 7. FML of this vintage scans every jar on the classpath for annotations it recognizes (such as Mod) and throws a noisy error if it encounters an issue doing so.

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

**Comment by exaskye** [comment]
It appears NewLaunch in PolyMC has gotten refactored, and makes use of lambdas in a few places.

It's understandable if this is a wontfix and I can try to implement a workaround, but I figured I should report it anyway.

**Comment by DioEgizio** [comment]
https://github.com/PolyMC/PolyMC/blob/8110040f869aa784065f184feaa339d73e11757e/libraries/launcher/CMakeLists.txt#L3

It's actually being built against java 7

**Comment by Scrumplex** [comment]
Nope. It targets Java 8

https://github.com/PolyMC/PolyMC/blob/8110040f869aa784065f184feaa339d73e11757e/libraries/launcher/CMakeLists.txt#L7

**Comment by DioEgizio** [comment]
oh nvm, it searches java 7 but builds with java 8

**Comment by DioEgizio** [comment]
> FAILED: jars/CMakeFiles/NewLaunch.dir/java_compiled_NewLaunch /home/runner/work/PolyMC/PolyMC/build/jars/CMakeFiles/NewLaunch.dir/java_compiled_NewLaunch 
cd /home/runner/work/PolyMC/PolyMC/libraries/launcher && /opt/hostedtoolcache/Java_Zulu_jdk/7.0.342-1/x64/bin/javac -target 7 -source 7 -Xlint:deprecation -Xlint:unchecked -classpath :/home/runner/work/PolyMC/PolyMC/libraries/launcher:/home/runner/work/PolyMC/PolyMC/build/jars -d /home/runner/work/PolyMC/PolyMC/build/jars/CMakeFiles/NewLaunch.dir org/multimc/EntryPoint.java org/multimc/Launcher.java org/multimc/LegacyFrame.java org/multimc/NotFoundException.java org/multimc/ParamBucket.java org/multimc/ParseException.java org/multimc/Utils.java org/multimc/onesix/OneSixLauncher.java net/minecraft/Launcher.java && /usr/local/bin/cmake -E touch /home/runner/work/PolyMC/PolyMC/build/jars/CMakeFiles/NewLaunch.dir/java_compiled_NewLaunch
org/multimc/ParamBucket.java:31: error: illegal start of expression
        paramsMap.computeIfAbsent(key, k -> new ArrayList<>())
                                          ^


ok no, building with java 7 doesn't work, probably the refactor broke it

**Comment by DioEgizio** [comment]
https://github.com/DioEgizio/PolyMC/actions/runs/2253216096 yeah works reverting that pr

**Comment by DioEgizio** [comment]
https://github.com/DioEgizio/PolyMC/actions/runs/2253261554 should work, now the question is if it's actually worth it to revert that refactor

**Comment by glowiak** [comment]
Works fine for me

**Comment by icelimetea** [comment]
Java 8 support was introduced in ASM 5.0 and it seems like Forge was updated to this ASM version in one of the releases for MC 1.7.10 (I have not checked does the same error appears on those newer versions, however)

> now the question is if it's actually worth it to revert that refactor

Actually Java 8 is not that essential for this refactor - it's possible to get rid of dependency on it's features (despite the fact that Java 8 features are nice to have).

**Comment by exaskye** [comment]
I'll note that updating ASM via a component causes the game to crash during launch because a check was added for the API version parameter, so this can't be fixed just by updating ASM without also patching mods to use the version param correctly.

**Comment by icelimetea** [comment]
Actually, it looks like that some ancient Forge versions don't launch on Java 8 at all, so maybe providing Java 7 support is a good idea.

**Comment by exaskye** [comment]
If the error is "Comparison method violates its general contract!" then you need to add `-Djava.util.Arrays.useLegacyMergeSort=true` to JVM arguments. I've not encountered any old Forge versions that can't run under Java 8, as long as that arg or LegacyJavaFixer (which does the same thing, but in a much hackier fashion) is present.

**Comment by icelimetea** [comment]
> I've not encountered any old Forge versions that can't run under Java 8, as long as that arg or LegacyJavaFixer (which does the same thing, but in a much hackier fashion) is present.

Hm, Forge 10.3.0.1208 for MC 1.7.10 doesn't work for me on Java 8. :/

**Comment by exaskye** [comment]
Ah, ConcurrentModificationException. Fun. Guess they tightened up those checks in Java 8.

Does in fact work under Java 7 in MultiMC; I've never tested things other than the last released version of Forge for a given version, as there's usually no reason to go older...

**Comment by icelimetea** [comment]
> I've never tested things other than the last released version of Forge for a given version, as there's usually no reason to go older...

Well, last released Forge for MC 1.7.2 doesn't quite work for me with Java 8 too.

**Comment by exaskye** [comment]
That's arguably an issue, since "early 1.7" and "late 1.7" are quite different play experiences.

**Comment by icelimetea** [comment]
Now Forge 1.4.7 launches without this error (at least for me).

**Comment by icelimetea** [comment]
Perhaps this issue can be considered resolved as NewLaunch compiles for Java 7 now.

**Comment by exaskye** [comment]
This is technically fixed, but now #690 has taken its place.
